### ISSUE TITLE
Listing 8.1 could not be run

### ISSUE THREAD

**Issue by Joshuacourse** [carl24k/fight-churn#34]
Hi Carl, 

I really like your book and enjoy the content that you put together. 

While running listing 8.1, I came across an error that I found not locating the cause. I was running the code in the Jupyter notebook environment. 

Below scripts are successfully run. 

run_listing(8,0)
run_churn_listing.run_listing(7,[3,4,6,8],insert=True)
run_churn_listing.run_listing(chapter=7,listing=[1],version=[1,2,3,4,5,6],insert=True)

Then when I ran "run_listing(8,0)", the below files are created and error message came,
Running chapter 8 listing 1 prepare_data on schema socialnet7
Saving results to output_jupyter/socialnet7/socialnet7_dataset2_summarystats.csv
Saving results to output_jupyter/socialnet7/socialnet7_dataset2_scores.csv
Saving params to output_jupyter/socialnet7/socialnet7_dataset2_score_params.csv

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/b7/2m41z5cd2p9c5qrqt0xrmz7c0000gn/T/ipykernel_1835/1734770269.py in <module>
      2 # run_churn_listing.run_listing(7,[3,4,6,8],insert=True) #
      3 # run_churn_listing.run_listing(chapter=7,listing=[1],version=[1,2,3,4,5,6],insert=True)
----> 4 run_listing(8,1)

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/run_churn_listing.py in run_listing(chapter, listing, version, schema, insert)
    338                      schema=schema,
    339                      insert=insert)
--> 340     run_churn_listing_from_args(args)
    341 
    342 

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/run_churn_listing.py in run_churn_listing_from_args(args)
    353         if len(args.version)==0:
    354             list_args.version=None
--> 355             run_listing_from_args(list_args)
    356         else:
    357             for v in args.version:

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/run_churn_listing.py in run_listing_from_args(args)
    287         _sql_listing_from_params(listing_params,args.insert)
    288     elif type=='py':
--> 289         _python_listing_from_params(listing_params)
    290     else:
    291         raise Exception('Unsupported type %s' % type)

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/run_churn_listing.py in _python_listing_from_params(param_dict)
    157     name = param_dict.pop('name')
    158 
--> 159     python_listing(chapter, listing, name, param_dict)
    160 
    161 

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/run_churn_listing.py in python_listing(chapter, listing, name, param_dict)
    189     if ex_fun is not None:
    190         # Run the function, passing the parameters
--> 191         ex_fun(**example_params)
    192     else:
    193         print('Could not find function %s in module %s' % (name , full_name))

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/listings/chap8/listing_8_1_prepare_data.py in prepare_data(data_set_path, group_corr_thresh)
      8     dataset_stats(data_set_path)
      9     fat_tail_scores(data_set_path)
---> 10     find_metric_groups(data_set_path,group_corr_thresh)
     11     apply_metric_groups(data_set_path)
     12     ordered_correlation_matrix(data_set_path)

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/listings/chap6/listing_6_4_find_metric_groups.py in find_metric_groups(data_set_path, group_corr_thresh)
     63     metric_columns = list(score_data.columns.values)
     64 
---> 65     labels = find_correlation_clusters(score_data.corr(),group_corr_thresh)
     66     labeled_column_df, relabled_count = relabel_clusters(labels,metric_columns)
     67     loadmat_df = make_load_matrix(labeled_column_df, metric_columns, relabled_count,group_corr_thresh)

~/Desktop/Life/DS Lab/Churn Modeling/fightchurn/listings/chap6/listing_6_4_find_metric_groups.py in find_correlation_clusters(corr, corr_thresh)
      9 def find_correlation_clusters(corr,corr_thresh):
     10     dissimilarity = 1.0 - corr
---> 11     hierarchy = linkage(squareform(dissimilarity), method='single')
     12     diss_thresh = 1.0 - corr_thresh
     13     labels = fcluster(hierarchy, diss_thresh, criterion='distance')

~/anaconda3/lib/python3.9/site-packages/scipy/spatial/distance.py in squareform(X, force, checks)
   2213             raise ValueError('The matrix argument must be square.')
   2214         if checks:
-> 2215             is_valid_dm(X, throw=True, name='X')
   2216 
   2217         # One-side of the dimensions is set here.

~/anaconda3/lib/python3.9/site-packages/scipy/spatial/distance.py in is_valid_dm(D, tol, throw, name, warning)
   2288             if not (D == D.T).all():
   2289                 if name:
-> 2290                     raise ValueError(('Distance matrix \'%s\' must be '
   2291                                      'symmetric.') % name)
   2292                 else:

ValueError: Distance matrix 'X' must be symmetric.

Just wondering if there is any dependency I missed. 

Looking forward to hearing from you!

Best
Josh

**Comment by carl24k** [comment]
This is interesting and I have not seen it before. It looks like some error came up in the clustering function. Can you share the correlation matrix you got from listing 6.2?

**Comment by Joshuacourse** [comment]
Sure. The correlation matrix is attached. 
[socialnet7_dataset_correlation_matrix.csv](https://github.com/carl24k/fight-churn/files/8792268/socialnet7_dataset_correlation_matrix.csv)

**Comment by Joshuacourse** [comment]
I did check the correlation matrix and there were a lot of NAN for multiple columns. I am not sure if I missed any prior steps, which is the dependency of listing 8.1. I am sharing with you the CSV file (outputted from score_data.corr() ) for the correlation matrix and dataset2.csv. For you to better understand how I came across the issue, I hence attach the PDF file reflecting the execution order of listings. All the listings for chapter 3,4,5,6,7 were executed properly and listing 8.1 starts giving the error message that is not covered by the book. 

Hope that helps! Thank you for your time and effort!
[correlation_metric.csv](https://github.com/carl24k/fight-churn/files/8792274/correlation_metric.csv)
[socialnet7_dataset2.csv](https://github.com/carl24k/fight-churn/files/8792275/socialnet7_dataset2.csv)
[Test.pdf](https://github.com/carl24k/fight-churn/files/8792276/Test.pdf)
